### ISSUE TITLE
Simplify WinRT and COM class authoring

### ISSUE THREAD

**Issue by kennykerr** [microsoft/windows-rs#1094]
Dedicating an issue to this topic. Originally part of #81 (huge thread).

The windows crate now supports implementing COM and WinRT interfaces, but more is required to support classes as as whole. This early sample illustrates what's possible today:

https://github.com/kennykerr/component-rs

I still have much work to do to streamline this experience.

**Comment by Alovchin91** [comment]
@kennykerr Is it possible today to author WinRT components in Rust, either using windows-rs crate or not? 🤔

I suppose it would require the use of Midl compiler + writing COM wiring manually. Could you maybe provide a sample of what can be done today?

Thanks!

**Comment by kennykerr** [comment]
You can do so today but yes you'd need to write some IDL and call the MIDL compiler to produce a winmd. Next year I'm hoping to spend most of my time on improving authoring support and being able to produce a winmd directly from Rust so that IDL will not be required.

**Comment by Alovchin91** [comment]
Thank you for the quick response! 😊 

I don't mind writing IDL and calling Midl manually. Having an ability to author WinRT components would unblock Rust's use for me 🙂

Do I understand correctly that I would need to call the `windows::build!` macro on the resulting WinMD, and then `#[implement]` interfaces from it? 🤔

**Comment by kennykerr** [comment]
Yep, that's it. It's a bit error-prone which is also why I'm planning on automating much of that.

**Comment by Alovchin91** [comment]
Yeah, I can imagine 😅  I'm gonna try this out and if I succeed, I could open a PR to the Samples repo, if you would be interested 🙂

**Comment by kennykerr** [comment]
Sounds good!

**Comment by Alovchin91** [comment]
Hi @kennykerr , you might be interested to take a look:

https://github.com/Alovchin91/winrt-component-rs

Please let me know if you have any suggestions, comments etc. 🙂

I'm not entirely sure about `*mut Option` in `DllGetActivationFactory` signature but that's the only way I managed to make it work.

**Comment by kennykerr** [comment]
Thanks @Alovchin91, that's very helpful. I'm now starting to work on this and #1093 in earnest. We should get to a point where MIDL is no longer required.

**Comment by hmyan90** [comment]
Synced a bit offline, just want to leave a comment here, please also consider prototype a winrt "event" type (similar like c++/winrt [winrt::event struct](https://docs.microsoft.com/en-us/uwp/cpp-ref-for-winrt/event) ) so when implement the event, we can know how to manage event handler and token. Thanks!

**Comment by kennykerr** [comment]
@hmyan90 take a look at #1705 - that should address your immediate need.

**Comment by kennykerr** [comment]
[0.36.0](https://github.com/microsoft/windows-rs/releases/tag/0.36.0) has been released and includes the new `Event<T>` type.

**Comment by nerocui** [comment]
Any update to this task? Looking forward to this so much!

**Comment by kennykerr** [comment]
I'm hard at work maturing support for component authoring. You can already implement components, with some limitations. There is an example here that you can use as a starting point:

https://github.com/microsoft/windows-rs/tree/master/crates/tests/component

**Comment by Ciantic** [comment]
Trying to understand this, I'm coming from [`com-rs` crate](https://github.com/microsoft/com-rs/), which was deprecated in favor of `windows-rs`, and it had this kind of macros:

`#[com_interface("a5cd92ff-29be-454c-8d04-d82879fb3f1b")]`
and
`#[co_class(implements(IVirtualDesktopNotification))]`

Is the test component example somehow the same thing but without macros? I can't find similar types of macro shortcuts in there, maybe it's more manual now?

---

## Usage examples

```rust
#[com_interface("a5cd92ff-29be-454c-8d04-d82879fb3f1b")]
pub trait IVirtualDesktopManager: IUnknown {
    unsafe fn is_window_on_current_desktop(
        &self,
        topLevelWindow: HWND,
        outOnCurrentDesktop: *mut bool,
    ) -> HRESULT;
    unsafe fn get_desktop_by_window(
        &self,
        topLevelWindow: HWND,
        outDesktopId: *mut DesktopID,
    ) -> HRESULT;
    unsafe fn move_window_to_desktop(
        &self,
        topLevelWindow: HWND,
        desktopId: *const DesktopID,
    ) -> HRESULT;
}
```
and

```rust
#[co_class(implements(IVirtualDesktopNotification))]
struct VirtualDesktopChangeListener {
    sender: Mutex<Option<VirtualDesktopEventSender>>,
}
```

Above are snipptes from my code  [I have a lots of code written](https://github.com/Ciantic/VirtualDesktopAccessor/blob/rust/src/interfaces.rs) with com-rs crate

**Comment by kennykerr** [comment]
I'm in the middle of developing first-class component authoring support, hence the lack of docs and samples but you can take [this example](https://github.com/microsoft/windows-rs/tree/master/crates/tests/component) as a guide. That happens to be a WinRT component but the same pattern applies for COM components. COM factories just implement `IClassFactory` instead of `IActivationFactory` and export `DllGetClassObject` instead of `DllGetActivationFactory`. 

By the way, your `IVirtualDesktopNotification` looks a lot like `IVirtualDesktopManager` which means you don't have to define it yourself and can just use the [definitions provided](https://microsoft.github.io/windows-docs-rs/doc/windows/?search=IVirtualDesktopManager) by the `windows` crate.

**Comment by Ciantic** [comment]
I got it working with similar macros: `windows_interface::interface` and `windows::core::implement`

What I don't get is this [instruction to use ManuallyDrop](https://github.com/microsoft/windows-rs/issues/1647#issuecomment-1085005439) whenever there is `_In_ IFooBar*` then use `ManuallyDrop<IFooBar>`... that didn't work for me, I just got a lot of problems that way.

```rust

#[windows_interface::interface("b2f925b9-5a0f-4d2e-9f4d-2b1507593c10")]
pub unsafe trait IVirtualDesktopManagerInternal: IUnknown {
    // Proc3
    unsafe fn get_count(&self, monitor: HMONITOR, outCount: *mut UINT) -> HRESULT;

    // Proc4
    unsafe fn move_view_to_desktop(
        &self,
        view: IApplicationView,
        desktop: IVirtualDesktop,
    ) -> HRESULT;

    // Proc5
    unsafe fn can_move_view_between_desktops(
        &self,
        view: IApplicationView,
        canMove: *mut i32,
    ) -> HRESULT;
}
```
And implementation:
```rust
#[windows::core::implement(IVirtualDesktopNotification)]
struct VirtualDesktopNotification {}

impl IVirtualDesktopNotification_Impl for VirtualDesktopNotification {
  unsafe fn virtual_desktop_created(
	  &self,
	  monitors: IObjectArray,
	  desktop: IVirtualDesktop,
  ) -> HRESULT {
	  HRESULT(0)
  }

  unsafe fn virtual_desktop_destroy_begin(
	  &self,
	  monitors: IObjectArray,
	  desktopDestroyed: IVirtualDesktop,
	  desktopFallback: IVirtualDesktop,
  ) -> HRESULT {
	  HRESULT(0)
  }
  // ....
}
```

There is also confusion should I use ManuallyDrop or not, someone said:

**Comment by Alovchin91** [comment]
You’re passing `IObjectArray` by value vs. by reference (`&IObjectArray`) which means you’re moving it into the function and the function now owns the value. When the function exits, it drops the value, which means the reference counter is decremented (`IUnknown.Release` is called). By using `ManuallyDrop<IObjectArray>` you’re essentially saying that the function should not drop the value. I believe an alternative option would be to `.clone()` your object array before passing it to the function — this way you get an `IUnknown.AddRef` call beforehand which is then matched by the `Release` call when the value is dropped.

**Comment by Ciantic** [comment]
@Alovchin91 Windows calls those functions, I give the instance to some register API.

If I've understood COM correctly the caller increments before calling, and the one using it releases at the end. So it should work without ManuallyDrops?

**Comment by tim-weis** [comment]
@Ciantic That seems to agree with how the [`[in]` attribute](https://learn.microsoft.com/en-us/windows/win32/Midl/in) is documented. If my understanding is correct, using the [`ManuallyDrop<T>`](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html) wrapper here would ultimately leak the object implementing the interface.

That leaves me wondering, though...

1. whether COM actually has an attribute to describe a "borrow"...
2. and if it doesn't whether [this table](https://github.com/microsoft/windows-rs/issues/1647#issuecomment-1085005439) is accurate.

**Comment by Ciantic** [comment]
I'm leaning that the table is not accurate. We should not use ManuallyDrop if COM API works as it should. However, a bigger example in the official FAQ might be in order.

The use case people need to see is translating C++ to Rust.

My current thought is this:

C++ `_In_ IApplicationView* pView` in windows-rs is `IApplicationView`

C++ `_In_opt_ HMONITOR monitor` in windows-rs is `Option<HMONITOR>`

C++ `_Out_ IVirtualDesktop**` in windows-rs is `*mut Option<IVirtualDesktop>`

<details>
E.g. here is C++

```cpp
MIDL_INTERFACE("B2F925B9-5A0F-4D2E-9F4D-2B1507593C10")
	IVirtualDesktopManagerInternal : public IUnknown
{
public:
	virtual HRESULT STDMETHODCALLTYPE GetCount(
		_In_opt_ HMONITOR monitor,
	_Out_ UINT* pCount) = 0;

	virtual HRESULT STDMETHODCALLTYPE MoveViewToDesktop(
		_In_ IApplicationView* pView,
		_In_ IVirtualDesktop* pDesktop) = 0;

	virtual HRESULT STDMETHODCALLTYPE CanViewMoveDesktops(
		_In_ IApplicationView* pView,
		_Out_ BOOL* pfCanViewMoveDesktops) = 0;

	virtual HRESULT STDMETHODCALLTYPE GetCurrentDesktop(
		_In_opt_ HMONITOR monitor,
		_Out_ IVirtualDesktop** desktop) = 0;

	virtual HRESULT STDMETHODCALLTYPE GetAllCurrentDesktops(
		_Out_ IObjectArray** ppDesktops) = 0;

	virtual HRESULT STDMETHODCALLTYPE GetDesktops(
		_In_opt_ HMONITOR monitor,
		_Out_ IObjectArray** ppDesktops) = 0;

	virtual HRESULT STDMETHODCALLTYPE GetAdjacentDesktop(
		_In_ IVirtualDesktop* pDesktopReference,
		_In_ AdjacentDesktop uDirection,
		_Out_ IVirtualDesktop** ppAdjacentDesktop) = 0;

	virtual HRESULT STDMETHODCALLTYPE SwitchDesktop(
		_In_opt_ HMONITOR monitor,
		_In_ IVirtualDesktop* pDesktop) = 0;

	virtual HRESULT STDMETHODCALLTYPE CreateDesktopW(
		_In_opt_ HMONITOR monitor,
		_Out_ IVirtualDesktop** ppNewDesktop) = 0;
        // ... 
}
```

I think it translates to this:

```rust
#[windows_interface::interface("b2f925b9-5a0f-4d2e-9f4d-2b1507593c10")]
pub unsafe trait IVirtualDesktopManagerInternal: IUnknown {
    unsafe fn get_count(&self, monitor: Option<HMONITOR>, outCount: *mut UINT) -> HRESULT;

    unsafe fn move_view_to_desktop(
        &self,
        view: IApplicationView,
        desktop: IVirtualDesktop,
    ) -> HRESULT;

    unsafe fn can_move_view_between_desktops(
        &self,
        view: IApplicationView,
        canMove: *mut i32,
    ) -> HRESULT;

    unsafe fn get_current_desktop(
        &self,
        monitor: HMONITOR,
        outDesktop: *mut Option<IVirtualDesktop>,
    ) -> HRESULT;

    unsafe fn get_all_current_desktops(&self, outDesktops: *mut Option<IObjectArray>) -> HRESULT;

    unsafe fn get_desktops(
        &self,
        monitor: HMONITOR,
        outDesktops: *mut Option<IObjectArray>,
    ) -> HRESULT;

    unsafe fn get_adjacent_desktop(
        &self,
        inDesktop: IVirtualDesktop,
        direction: UINT,
        out_pp_desktop: *mut Option<IVirtualDesktop>,
    ) -> HRESULT;

    unsafe fn switch_desktop(&self, monitor: HMONITOR, desktop: IVirtualDesktop) -> HRESULT;

    unsafe fn create_desktop(
        &self,
        monitor: HMONITOR,
        outDesktop: *mut Option<IVirtualDesktop>,
    ) -> HRESULT;
    // ...
}
```



</details>

**Comment by kennykerr** [comment]
This is slightly confusing as a COM interface pointer is modeled as a value type in Rust. If you think of it in terms of C++ it makes a little more sense conceptually. An input parameter passes the raw pointer into the function. The caller ensures that the pointer is stable for the duration of the synchronous call and the callee depends on that assurance, but the caller does not transfer ownership to the callee. Rust models it more like a C++ smart pointer, but such abstractions are not valid on the ABI where the parameter must ultimately be the equivalent of a raw C++ pointer. This is why @wesleywiser correctly suggests using `ManuallyDrop`. 

Anyway, I still plan to make this a lot simpler and safer in Rust. If you want to understand how this all works, I explain all of this and much more in great detail here:

https://www.pluralsight.com/courses/com-essentials

**Comment by Ciantic** [comment]
> The caller ensures that the pointer is stable for the duration of the synchronous call and the callee depends on that assurance, but the caller does not transfer ownership to the callee.

This does sound like `clone()` before calling or `ManuallyDrop` is required.

I have used [ComPtr in C++](https://learn.microsoft.com/en-us/cpp/cppcx/wrl/comptr-class?view=msvc-170) successfully, and ComRc and ComPtr in `com-rs` crate. I guess something like that would be nice in here too, now this feels pretty error-prone, but I think if I can remember to clone each time this works.

I actually have the book [Essential COM, Don Box](https://www.amazon.com/Essential-COM-Don-Box/dp/0201634465) it feels like I have enough COM info for my lifetime, maybe your content is more succinct.

**Comment by kennykerr** [comment]
Keep in mind that COM relies on the `stdcall` calling convention (or 64-bit equivalents) which requires the caller to pack the stack and the callee to unpack the stack. What that means is that if the caller passes an object with destructor, such as a `Drop` implementation, by value then the compiler will assume the callee will either assume ownership or drop the value. But that is not what COM expects so if you're doing that on your end (e.g. passing a cloned value) you're going to cause a COM reference leak when the callee is written in something other than Rust.

**Comment by Ciantic** [comment]
Now I'm just getting clever.

```cpp
virtual HRESULT STDMETHODCALLTYPE SwitchDesktop(
    _In_opt_ HMONITOR monitor,
    _In_ IVirtualDesktop* pDesktop) = 0;
```

For that `_In_` I made this


```rust
#[repr(C)]
pub struct ComIn<T>(ManuallyDrop<T>);
impl<T: Clone> ComIn<T> {
    fn new(t: &T) -> Self {
        Self(ManuallyDrop::new(t.clone()))
    }
}
```

And use it in interface definitions like this:

```rust
    unsafe fn switch_desktop(&self, monitor: HMONITOR, desktop: ComIn<IVirtualDesktop>) -> HRESULT;
```

And call it like this: 

```rust
unsafe {
    manager
        .switch_desktop(0, ComIn::new(&current_desk))
        .unwrap()
};
```

**Comment by kennykerr** [comment]
You should be able to implement IDispatch in this scenario. Be sure to include the necessary feature requirements:

```
[dependencies.windows]
version = "0.46.0"
features = [
    "implement",
    "Win32_Foundation",
    "Win32_System_Com",
    "Win32_System_Ole",
]
```

Then you just need to include an implementation:

```rust
impl IDispatch_Impl for HostObjectSample {
    fn GetTypeInfoCount(&self) -> Result<u32> { todo!() }
    fn GetTypeInfo(&self, _: u32, _: u32) -> Result<ITypeInfo> { todo!() }
    fn GetIDsOfNames(&self, _: *const GUID, _: *const PCWSTR, _: u32, _: u32, _: *mut i32) -> Result<()> { todo!() }
    fn Invoke(&self, _: i32, _: *const GUID, _: u32, _: DISPATCH_FLAGS, _: *const DISPPARAMS, _: *mut VARIANT, _: *mut EXCEPINFO, _: *mut u32) -> Result<()> { todo!() }
}
```

**Comment by defims** [comment]
@kennykerr would you please provide a simple example? I found a [c++ version](https://stackoverflow.com/questions/3747414/calling-c-function-from-javascript-script-running-in-a-web-browser-control) in stackoverflow, but a simpler example of Rust version would be great.

**Comment by lesderid** [comment]
For WebView2 host objects, you can skip implementing `GetTypeInfo` and just return `Ok(0)` from `GetTypeInfoCount`. This only leaves implementing `GetIDsOfNames` and `Invoke`, which is trivial for most cases, only the `VARIANT`s are a bit annoying.

**Comment by defims** [comment]
I wrote a [macro](https://crates.io/crates/wvwasi-macro) called `wvwasi_macro::create_type_info_crate` that automatically generates a trait for getting `ITypeInfo`, which simplifies the use of `AddHostObjectToScript`. Welcome everyone to use it.

**Comment by Ciantic** [comment]
Something changed between 0.53 and 0.56, this code

```
#[interface("094d70d6-5202-44b8-abb8-43860da5aca2")]
unsafe trait IValue: IUnknown {
    fn GetValue(&self, value: *mut i32) -> HRESULT;
}
```

Began to give an error that it requires a "windows_core" crate, I added `windows-core` crate and it works again.

**Comment by kennykerr** [comment]
https://github.com/microsoft/windows-rs/pull/2917

**Comment by Alovchin91** [comment]
#3266 

> In an attempt to streamline and focus on what's most valuable, I'm making the difficult decision to drop support for metadata generation in general. While the existing metadata generation tooling is not ideal for Rust developers (e.g. midlrt and win32metadata), I don't realistically have the bandwidth to provide a Rust version any time soon.

I think this actually makes a lot of sense, as long as authoring WinRT components is still a supported scenario. IDL works well and aligns with what C++/WinRT developers do. If we could get metadata generation support via IDL with shim like DllGetActivationFactory, DllCanUnloadNow etc. generation instead, that would be awesome 👍

**Comment by Alovchin91** [comment]
> DllGetActivationFactory

Alternatively, having a generated associated `GetRuntimeClassName` function on authored components would’ve already been a huge improvement for manual authoring support.

**Comment by Alovchin91** [comment]
> What about using the `RuntimeName` trait?

Oh! I didn't realise it was available. Great, thanks! 👍

**Comment by Alovchin91** [comment]
@kennykerr Just checking if I've missed it: is there now a reference counter for `DllCanUnloadNow`, as mentioned in https://github.com/microsoft/windows-rs/issues/2472#issuecomment-1520090678?

**Comment by kennykerr** [comment]
This isn't directly supported and I still stand by this comment: 

https://github.com/microsoft/windows-rs/issues/2472#issuecomment-1520618377

It's just not a good idea or practically impossible to implement reliably in a non-trivial system. It also adds measurable run time cost and complexity as it interacts with things like factory caching and every single implementation path.

**Comment by Alovchin91** [comment]
Okay, so I didn't miss anything 😄 Sounds fair to me, thanks!
